Update docstring tag pep and grammar (#1693)
python-discord/bot: bot/resources/tags/docstring.md

> A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string with triple quotes that's placed at the top of files, classes and functions. A docstring should contain a clear explanation of what it's describing. You can also include descriptions of the subject's parameter(s) and what it returns, as shown below:

Reviewer: ```suggestion
A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string (always using triple quotes) that's placed at the top of files, classes and functions. A docstring should contain a clear explanation of what it's describing. You can also include descriptions of the subject's parameter(s) and what it returns, as shown below:
```

> A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string with triple quotes that's placed at the top of files, classes and functions. A docstring should contain a clear explanation of what it's describing. You can also include descriptions of the subjects's parameter(s) and its return type, as shown below.

Reviewer: > A docstring should contain a clear explanation of what it's describing.

Huh?

Author: We didn't want to enumerate files, classes and functions again, so we went with this instead, not sure how to improve this.

Reviewer: What about..
> A docstring should contain a clear explanation of what it does and its purpose.

Author: it would need to be referring to the subject here, rather than the docstring itself. So what about `A docstring should contain a clear explanation of what it's describing does and its purpose.`

Reviewer: `what it's describing does` still rubs me the wrong way.. what if we change that to `whatever`?

Or I guess completely different more like `A docstring should be a clear explanation of the purpose of what it describes.`

Reviewer 2: I think something along the lines of `A docstring should describe the behaviour and the interface of what it's attached to` would fit.

Reviewer: I like `what it's attached to`, but a docstring should not explain the interface. That is up to the function signature to do.

Reviewer 2: With the interface I meant describing what some of the more complex parameters mean, not just how the signature looks- - maybe the original wording is clearer there

> A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string, always using triple quotes, that's placed at the top of files, classes and functions. A docstring should contain a clear explanation of what it's describing. You can also include descriptions of the subject's parameter(s) and what it returns, as shown below:

Reviewer: We could probably uses dashes for the triple quote comment?
```suggestion
A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string - always using triple quotes - that's placed at the top of files, classes and functions. A docstring should contain a clear explanation of what it's describing. You can also include descriptions of the subject's parameter(s) and what it returns, as shown below:
```

```py
def greet(name: str, age: int) -> str:
    """
    Return a string that greets the given person, including their name and age.

    :param name: The name to greet.

>     :param age: The age to display.

Reviewer: ```suggestion
    :param age: The person's age.
```

Author: I was actually about to commit `The age of the person to include in the greeting.` What do you think?

Reviewer: Something more like `The age of the person to greet`? Since we want to know the age of the person that we're about to greet, the person isn't necessarily *included*.

>     :return: String representation of the greeting.

Reviewer: We should be consistent here, either change all of them or change this to..
```suggestion
    :return: The string representation of the greeting.
```

    """
    return f"Hello {name}, you are {age} years old!"
```

> You can get the docstring by using the `.__doc__` attribute. For the last example, you can print it by doing this: `print(greet.__doc__)`.

Reviewer 2: Would `inspect.getdoc` make more sense here? I'm not sure if the attribute is intended to be accessed directly, or something like getdoc which also cleans it up is a better alternative

Reviewer: It's kind of like `dir()` vs. `.__dict__`.

Author: yea, good idea. Added in https://github.com/python-discord/bot/pull/1693/commits/39500b40ac2892cb469366103022bd854f7734b2

Author: > It's kind of like `dir()` vs. `.__dict__`.

inspect.getdoc Also clears indents, which is nice

> For more details about what a docstring is and its usage, check out this guide by [Real Python](https://realpython.com/documenting-python-code/#docstrings-background), or the [PEP-257 docs](https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring).

Reviewer: I kind of want to include some form of note that explains that a PEP is the official specification/proposal for those that don't know.

Maybe just?
```suggestion
For more details about what a docstring is and its usage, check out this guide by [Real Python](https://realpython.com/documenting-python-code/#docstrings-background), or the [official docstring specification](https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring).
```

> A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string with triple quotes that's often used in files, classes, functions, etc. A docstring should contain a clear explanation of exactly what the function does. You can also include descriptions of the function's parameter(s) and its return type, as shown below.

Reviewer: Why have the `etc.`? That's all the uses..
```suggestion
A [`docstring`](https://docs.python.org/3/glossary.html#term-docstring) is a string with triple quotes that's often at the top of files, classes, and functions. A docstring should contain a clear explanation of exactly what the function does. You can also include descriptions of the function's parameter(s) and its return type, as shown below.
```

```py
def greet(name, age) -> str:

>     """

Reviewer: Doesn't PEP 257 say docstrings should be written like this:
```py
"""A magic method, we don't know what it does.

Only kept here for legacy reasons, if we remove it the code breaks.
"""
```

Author: I don't think it mandates that, it uses that in some examples, but then also uses the blank first line as an example lower down, saying they're equivalent. At least that's how I read the pep

    Return a string that greets the given person, including their name and age.

    :param name: The name to greet.
    :type name: str
    :param age: The age to display.
    :type age: int

    :return: String representation of the greeting.
    """

>     return_string = f"Hello, {name} you are {age} years old!"

Reviewer: Unnecessary variable?

    return return_string
```

> You can get the docstring by using the `.__doc__` attribute. For the last example you can get it through: `print(greet.__doc__)`.

Reviewer: ```suggestion
You can get the docstring by using the `.__doc__` attribute. For the last example you can print it like this: `print(greet.__doc__)`.
```
